### ISSUE TITLE
Some questions on best practices (implementing a multivariate normal module)

### ISSUE THREAD

**Issue by bayerj** [google/flax#274]
Hello,

I am currently evaluating flax in a context where some components are parameterized, but don't necessarily have an obvious "output" of some kind. 

E.g. we might consider implementing a multivariate normal (MVN). There are all kinds of operations that I want to perform here: sampling, calculating the entropy, calculating the kl divergence from a different MVN and of course calculating the log probability of an observation.

Given the downstream use, I'd want those to be trained through a loss function as well. Now, here is a quick (functional) sketch how I could solve that with multi methods:

```python
import flax
import flax.nn
import jax
import jax.random
import jax.numpy as jnp
import jax.nn.initializers
import jax.scipy.stats as jst


class MVND(flax.nn.Module):
    def apply(self, dim):
        self._loc(dim)
        self._scale_diag(dim)

    def _loc(self, dim):
        return self.param("loc", (dim,), jax.nn.initializers.zeros)

    def _scale_diag(self, dim):
        unrect_scale_diag = self.param(
            "unrect_scale_diag", (dim,), jax.nn.initializers.zeros
        )
        return jnp.exp(unrect_scale_diag)

    @flax.nn.module_method
    def log_prob(self, inpt, dim):
        return jst.norm.logpdf(
            inpt, loc=self._loc(dim), scale=self._scale_diag(dim)
        ).sum(-1)


mvnd = MVND.partial(dim=1)
_, params = mvnd.init(jax.random.PRNGKey(32))
model = flax.nn.Model(mvnd, params)

print(model.log_prob(jnp.zeros((2, 1))))
```

I have a few questions on whether this is idiomatic flax.

1. The `apply` method does not return anything. That is because there is no obvious return value, as no real "forward" pass exists here. It somehow degenerates into a parameter declaration method. Is this how it should be done?
2. The `log_prob` calls `_loc` and `_scale_diag`. The issue here is that arguments (`dim` in this case) need to be propagated explicitly. This gives me a feeling that this approach won't scale gracefully, as the number of such arguments can easily grow. Still, these are  constant over the lifetime of a single parameter set or `Model` instance later on. Is there a way to keep track of those? In an OOP world, these would correspond to instance level attributes.

Thanks for any help!

**Comment by avital** [comment]
Hi @bayerj!

Indeed you've found some of the less intuitive parts of the current Flax API, in particular using apply as just a parameter-declaration method, and the hard-to-discover pattern where you call methods like `_scale_diag` and `_loc` to define parameters being used from other module methods. We hope that #208 (still being discussed) could go some way towards a simpler mental model for examples like these.

To answer you questions concretely,
1. Yes, that's the best way to do this right now.
2. If I understand you correctly, indeed the clearer pattern here would be to store them as attributes on the module instance and then use them from all of the relevant methods. That would also be made much simpler with #208. 

What do you think? Would this resolve your concerns?

**Comment by avital** [comment]
And please let us know what else you find confusing about Flax! (Perhaps in another issue)

**Comment by bayerj** [comment]
Thanks for the answer and the pointer. I will check out that issue.

**Comment by marcvanzee** [comment]
Closing this for now since this issue seems to be resolved, but feel free to re-open or open another issue if something is still unclear;
